File: flutter_secure_storage/ios/Classes/FlutterSecureStoragePlugin.m
#import "FlutterSecureStoragePlugin.h"
#if __has_include(<flutter_secure_storage/flutter_secure_storage-Swift.h>)
#import <flutter_secure_storage/flutter_secure_storage-Swift.h>
#else
// Support project import fallback if the generated compatibility header
// is not copied when this plugin is created as a library.
// https://forums.swift.org/t/swift-static-libraries-dont-copy-generated-objective-c-header/19816
#import "flutter_secure_storage-Swift.h"
#endif

@implementation FlutterSecureStoragePlugin
+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  [SwiftFlutterSecureStoragePlugin registerWithRegistrar:registrar];
}
@end


//#import "FlutterSecureStoragePlugin.h"
//
//static NSString *const CHANNEL_NAME = @"plugins.it_nomads.com/flutter_secure_storage";
//
//static NSString *const InvalidParameters = @"Invalid parameter's type";
//
//@interface FlutterSecureStoragePlugin()
//
//@property (strong, nonatomic) NSDictionary *query;
//
//@end
//
//@implementation FlutterSecureStoragePlugin
//
//- (instancetype)init {
//    self = [super init];
//    if (self){
//        self.query = @{
//                       (__bridge id)kSecClass :(__bridge id)kSecClassGenericPassword,
//                       };
//    }
//    return self;
//}
//
//+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
//    FlutterMethodChannel* channel = [FlutterMethodChannel
//                                     methodChannelWithName:CHANNEL_NAME
//                                     binaryMessenger:[registrar messenger]];
//    FlutterSecureStoragePlugin* instance = [[FlutterSecureStoragePlugin alloc] init];
//    [registrar addMethodCallDelegate:instance channel:channel];
//}
//
//- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
//    NSDictionary *arguments = [call arguments];
//    NSDictionary *options = [arguments[@"options"] isKindOfClass:[NSDictionary class]] ? arguments[@"options"] : nil;
//    NSString *accountName = options[@"accountName"];
//    NSString *groupId = options[@"groupId"];
//    NSString *synchronizable = options[@"synchronizable"];
//
//    if ([@"read" isEqualToString:call.method]) {
//        NSString *key = arguments[@"key"];
//        NSString *value = [self read:key forGroup:groupId forAccountName:accountName forSynchronizable:synchronizable];
//
//        result(value);
//    } else if ([@"write" isEqualToString:call.method]) {
//        NSString *key = arguments[@"key"];
//        NSString *value = arguments[@"value"];
//        NSString *accessibility = options[@"accessibility"];
//
//        if (![value isKindOfClass:[NSString class]]){
//            result(InvalidParameters);
//            return;
//        }
//
//        [self write:value forKey:key forGroup:groupId accessibilityAttr:accessibility forAccountName:accountName forSynchronizable:synchronizable];
//
//        result(nil);
//    } else if ([@"delete" isEqualToString:call.method]) {
//        NSString *key = arguments[@"key"];
//
//        [self delete:key forGroup:groupId forAccountName:accountName forSynchronizable:synchronizable];
//
//        result(nil);
//    } else if ([@"deleteAll" isEqualToString:call.method]) {
//        [self deleteAll: groupId forAccountName:accountName forSynchronizable:synchronizable];
//
//        result(nil);
//    } else if ([@"readAll" isEqualToString:call.method]) {
//        NSDictionary *value = [self readAll: groupId forAccountName:accountName forSynchronizable:synchronizable];
//
//        result(value);
//    } else if ([@"containsKey" isEqualToString:call.method]) {
//        NSString *key = arguments[@"key"];
//        NSNumber *containsKey = [self containsKey:key forGroup:groupId forAccountName:accountName forSynchronizable:synchronizable];
//
//        result(containsKey);
//    } else {
//        result(FlutterMethodNotImplemented);
//    }
//}
//
//- (void)write:(NSString *)value forKey:(NSString *)key forGroup:(NSString *)groupId accessibilityAttr:(NSString *)accessibility forAccountName:(NSString *)accountName forSynchronizable:(NSString *)synchronizable {
//    NSMutableDictionary *search = [self.query mutableCopy];
//
//    if(groupId != nil) {
//        search[(__bridge id)kSecAttrAccessGroup] = groupId;
//    }
//
//    if(accountName != nil) {
//        search[(__bridge id)kSecAttrService] = accountName;
//    }
//
//    search[(__bridge id)kSecAttrAccount] = key;
//    search[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitOne;
//
//    CFBooleanRef attrSynchronizable = kCFBooleanFalse;
//    if ([synchronizable isEqualToString:@"true"]) {
//        attrSynchronizable = kCFBooleanTrue;
//    } else {
//        attrSynchronizable = kCFBooleanFalse;
//    }
//    search[(__bridge id)kSecAttrSynchronizable] = (__bridge id) attrSynchronizable;
//
//    // The default setting is kSecAttrAccessibleWhenUnlocked
//    CFStringRef attrAccessible = kSecAttrAccessibleWhenUnlocked;
//    if (accessibility != nil) {
//        if ([accessibility isEqualToString:@"passcode"]) {
//            attrAccessible = kSecAttrAccessibleWhenPasscodeSetThisDeviceOnly;
//        } else if ([accessibility isEqualToString:@"unlocked"]) {
//            attrAccessible = kSecAttrAccessibleWhenUnlocked;
//        } else if ([accessibility isEqualToString:@"unlocked_this_device"]) {
//            attrAccessible = kSecAttrAccessibleWhenUnlockedThisDeviceOnly;
//        } else if ([accessibility isEqualToString:@"first_unlock"]) {
//            attrAccessible = kSecAttrAccessibleAfterFirstUnlock;
//        } else if ([accessibility isEqualToString:@"first_unlock_this_device"]) {
//            attrAccessible = kSecAttrAccessibleAfterFirstUnlockThisDeviceOnly;
//        }
//    }
//
//    OSStatus status;
//    status = SecItemCopyMatching((__bridge CFDictionaryRef)search, NULL);
//    if (status == noErr){
//        search[(__bridge id)kSecMatchLimit] = nil;
//
//        NSDictionary *update = @{
//            (__bridge id)kSecValueData: [value dataUsingEncoding:NSUTF8StringEncoding],
//            (__bridge id)kSecAttrAccessible: (__bridge id) attrAccessible,
//            (__bridge id)kSecAttrSynchronizable: (__bridge id) attrSynchronizable,
//        };
//
//        status = SecItemUpdate((__bridge CFDictionaryRef)search, (__bridge CFDictionaryRef)update);
//        if (status != noErr){
//            NSLog(@"SecItemUpdate status = %d", (int) status);
//        }
//    }else{
//        search[(__bridge id)kSecValueData] = [value dataUsingEncoding:NSUTF8StringEncoding];
//        search[(__bridge id)kSecMatchLimit] = nil;
//        search[(__bridge id)kSecAttrAccessible] = (__bridge id) attrAccessible;
//
//        status = SecItemAdd((__bridge CFDictionaryRef)search, NULL);
//        if (status != noErr){
//            NSLog(@"SecItemAdd status = %d", (int) status);
//        }
//    }
//}
//
//- (NSString *)read:(NSString *)key forGroup:(NSString *)groupId forAccountName:(NSString *)accountName forSynchronizable:(NSString *)synchronizable {
//    NSMutableDictionary *search = [self.query mutableCopy];
//    if(groupId != nil) {
//     search[(__bridge id)kSecAttrAccessGroup] = groupId;
//    }
//    if(accountName != nil) {
//        search[(__bridge id)kSecAttrService] = accountName;
//    }
//    search[(__bridge id)kSecAttrAccount] = key;
//    search[(__bridge id)kSecReturnData] = (__bridge id)kCFBooleanTrue;
//
//    if ([synchronizable isEqualToString:@"true"]) {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanTrue;
//    } else {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanFalse;
//    }
//
//    CFDataRef resultData = NULL;
//
//    OSStatus status;
//    status = SecItemCopyMatching((__bridge CFDictionaryRef)search, (CFTypeRef*)&resultData);
//    NSString *value;
//    if (status == noErr){
//        NSData *data = (__bridge NSData*)resultData;
//        value = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
//        CFRelease(resultData);
//    }
//    return value;
//}
//
//- (void)delete:(NSString *)key forGroup:(NSString *)groupId forAccountName:(NSString *)accountName forSynchronizable:(NSString *)synchronizable {
//    NSMutableDictionary *search = [self.query mutableCopy];
//    if(groupId != nil) {
//        search[(__bridge id)kSecAttrAccessGroup] = groupId;
//    }
//    if(accountName != nil) {
//        search[(__bridge id)kSecAttrService] = accountName;
//    }
//    search[(__bridge id)kSecAttrAccount] = key;
//    search[(__bridge id)kSecReturnData] = (__bridge id)kCFBooleanTrue;
//
//    if ([synchronizable isEqualToString:@"true"]) {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanTrue;
//    } else {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanFalse;
//    }
//
//    OSStatus status;
//    status = SecItemDelete((__bridge CFDictionaryRef)search);
//    if (status != noErr){
//        NSLog(@"SecItemDelete status = %d", (int) status);
//    }
//}
//
//- (void)deleteAll:(NSString *)groupId forAccountName:(NSString *)accountName forSynchronizable:(NSString *)synchronizable {
//    NSMutableDictionary *search = [self.query mutableCopy];
//    if(groupId != nil) {
//        search[(__bridge id)kSecAttrAccessGroup] = groupId;
//    }
//    if(accountName != nil) {
//        search[(__bridge id)kSecAttrService] = accountName;
//    }
//    if ([synchronizable isEqualToString:@"true"]) {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanTrue;
//    } else {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanFalse;
//    }
//    OSStatus status;
//    status = SecItemDelete((__bridge CFDictionaryRef)search);
//    if (status != noErr){
//        NSLog(@"SecItemDeleteAll status = %d", (int) status);
//    }
//}
//
//- (NSDictionary *)readAll:(NSString *)groupId forAccountName:(NSString *)accountName forSynchronizable:(NSString *)synchronizable {
//    NSMutableDictionary *search = [self.query mutableCopy];
//    if(groupId != nil) {
//        search[(__bridge id)kSecAttrAccessGroup] = groupId;
//    }
//    if(accountName != nil) {
//        search[(__bridge id)kSecAttrService] = accountName;
//    }
//
//    search[(__bridge id)kSecReturnData] = (__bridge id)kCFBooleanTrue;
//
//    search[(__bridge id)kSecMatchLimit] = (__bridge id)kSecMatchLimitAll;
//    search[(__bridge id)kSecReturnAttributes] = (__bridge id)kCFBooleanTrue;
//
//    if ([synchronizable isEqualToString:@"true"]) {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanTrue;
//    } else {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanFalse;
//    }
//
//    CFArrayRef resultData = NULL;
//
//    OSStatus status;
//    status = SecItemCopyMatching((__bridge CFDictionaryRef)search, (CFTypeRef*)&resultData);
//    if (status == noErr){
//        NSArray *items = (__bridge NSArray*)resultData;
//
//        NSMutableDictionary *results = [[NSMutableDictionary alloc] init];
//        for (NSDictionary *item in items){
//            NSString *key = item[(__bridge NSString *)kSecAttrAccount];
//            NSString *value = [[NSString alloc] initWithData:item[(__bridge NSString *)kSecValueData] encoding:NSUTF8StringEncoding];
//            results[key] = value;
//        }
//        CFRelease(resultData);
//        return [results copy];
//    }
//    return @{};
//}
//
//- (NSNumber *)containsKey:(NSString *)key forGroup:(NSString *)groupId forAccountName:(NSString *)accountName forSynchronizable:(NSString *)synchronizable {
//    NSMutableDictionary *search = [self.query mutableCopy];
//    if(groupId != nil) {
//        search[(__bridge id)kSecAttrAccessGroup] = groupId;
//    }
//    if(accountName != nil) {
//        search[(__bridge id)kSecAttrService] = accountName;
//    }
//    search[(__bridge id)kSecAttrAccount] = key;
//    search[(__bridge id)kSecReturnData] = (__bridge id)kCFBooleanTrue;
//
//    if ([synchronizable isEqualToString:@"true"]) {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanTrue;
//    } else {
//        search[(__bridge id)kSecAttrSynchronizable] = (__bridge id)kCFBooleanFalse;
//    }
//
//    if ([search objectForKey:(__bridge id)(kSecAttrAccount)]) {
//        return @YES;
//    } else {
//        return @NO;
//    }
//}
//
//@end

File: flutter_secure_storage_macos/macos/Classes/FlutterSecureStoragePlugin.swift
//
//  FlutterSecureStoragePlugin.swift
//  flutter_secure_storage_macos
//
//  Created by Julian Steenbakker on 09/12/2022.
//

import FlutterMacOS

public class FlutterSecureStoragePlugin: NSObject, FlutterPlugin {
    
    private let flutterSecureStorageManager: FlutterSecureStorage = FlutterSecureStorage()
    
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "plugins.it_nomads.com/flutter_secure_storage", binaryMessenger: registrar.messenger)
        let instance = FlutterSecureStoragePlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch (call.method) {
        case "read":
            read(call, result)
        case "write":
            write(call, result)
        case "delete":
            delete(call, result)
        case "deleteAll":
            deleteAll(call, result)
        case "readAll":
            readAll(call, result)
        case "containsKey":
            containsKey(call, result)
        case "isProtectedDataAvailable":
            // NSApplication is not thread safe
            DispatchQueue.main.async {
                if #available(macOS 12.0, *) {
                    result(NSApplication.shared.isProtectedDataAvailable)
                } else {
                    result(true)
                }
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    private func read(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let values = parseCall(call)
        if (values.key == nil) {
            result(FlutterError.init(code: "Missing Parameter", message: "write requires key parameter", details: nil))
            return
        }
        
        let response = flutterSecureStorageManager.read(key: values.key!, groupId: values.groupId, accountName: values.accountName, synchronizable: values.synchronizable, accessibility: values.accessibility)
        handleResponse(response, result)
    }
    
    private func write(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        if (!((call.arguments as! [String : Any?])["value"] is String)){
            result(FlutterError.init(code: "Invalid Parameter", message: "key parameter must be String", details: nil))
            return;
        }
        
        let values = parseCall(call)
        if (values.key == nil) {
            result(FlutterError.init(code: "Missing Parameter", message: "write requires key parameter", details: nil))
            return
        }
        
        if (values.value == nil) {
            result(FlutterError.init(code: "Missing Parameter", message: "write requires value parameter", details: nil))
            return
        }
        
        let response = flutterSecureStorageManager.write(key: values.key!, value: values.value!, groupId: values.groupId, accountName: values.accountName, synchronizable: values.synchronizable, accessibility: values.accessibility)
        
        handleResponse(response, result)
    }
    
    private func delete(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let values = parseCall(call)
        if (values.key == nil) {
            result(FlutterError.init(code: "Missing Parameter", message: "delete requires key parameter", details: nil))
            return
        }
        
        let response = flutterSecureStorageManager.delete(key: values.key!, groupId: values.groupId, accountName: values.accountName, synchronizable: values.synchronizable, accessibility: values.accessibility)
        
        handleResponse(response, result)
    }
    
    private func deleteAll(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let values = parseCall(call)
        let response = flutterSecureStorageManager.deleteAll(groupId: values.groupId, accountName: values.accountName, synchronizable: values.synchronizable)
        
        handleResponse(response, result)
    }
    
    private func readAll(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let values = parseCall(call)
        let response = flutterSecureStorageManager.readAll(groupId: values.groupId, accountName: values.accountName, synchronizable: values.synchronizable)
        
        handleResponse(response, result)
    }
    
    private func containsKey(_ call: FlutterMethodCall, _ result: @escaping FlutterResult) {
        let values = parseCall(call)
        if (values.key == nil) {
            result(FlutterError.init(code: "Missing Parameter", message: "containsKey requires key parameter", details: nil))
        }
        
        let response = flutterSecureStorageManager.containsKey(key: values.key!, groupId: values.groupId, accountName: values.accountName, synchronizable: values.synchronizable, accessibility: values.accessibility)
        
        switch response {
        case .success(let exists):
            result(exists)
            break;
        case .failure(let err):
            var errorMessage = ""

            if let errMsg = SecCopyErrorMessageString(err.status, nil) as? String {
                errorMessage = "Code: \(err.status), Message: \(errMsg)"
            } else {
                errorMessage = "Unknown security result code: \(err.status)"
            }
            result(FlutterError.init(code: "Unexpected security result code", message: errorMessage, details: err.status))
            break;
        }
    }
    
    private func parseCall(_ call: FlutterMethodCall) -> FlutterSecureStorageRequest {
        let arguments = call.arguments as! [String : Any?]
        let options = arguments["options"] as! [String : Any?]
        
        let accountName = options["accountName"] as? String
        let groupId = options["groupId"] as? String
        let synchronizableString = options["synchronizable"] as? String
        
        let synchronizable: Bool = synchronizableString != nil ? Bool(synchronizableString!)! : false
        
        let key = arguments["key"] as? String
        let accessibility = options["accessibility"] as? String
        let value = arguments["value"] as? String
        
        return FlutterSecureStorageRequest(
            accountName: accountName,
            groupId: groupId,
            synchronizable: synchronizable,
            accessibility: accessibility, 
            key: key, 
            value: value
        )
    }

    private func handleResponse(_ response: FlutterSecureStorageResponse, _ result: @escaping FlutterResult) {
        if let status = response.status {
            if (status == noErr) {
                result(response.value)
            } else {
                var errorMessage = ""

                if let errMsg = SecCopyErrorMessageString(status, nil) as? String {
                    errorMessage = "Code: \(status), Message: \(errMsg)"
                } else {
                    errorMessage = "Unknown security result code: \(status)"
                }
                result(FlutterError.init(code: "Unexpected security result code", message: errorMessage, details: status))
            }
        } else {
            result(response.value)
        }
    }
    
    struct FlutterSecureStorageRequest {
        var accountName: String?
        var groupId: String?
        var synchronizable: Bool?
        var accessibility: String?
        var key: String?
        var value: String?
    }
    
}
